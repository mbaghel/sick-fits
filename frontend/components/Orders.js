import React from "react";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import { formatDistance } from "date-fns";
import Link from "next/link";
import styled from "styled-components";
import OrderItemStyles from "./styles/OrderItemStyles";
import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";

const USER_ORDERS_QUERY = gql`
  query {
    orders {
      id
      items {
        id
        title
        image
        quantity
      }
      total
      createdAt
    }
  }
`;

const OrderUl = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

const Orders = () => {
  return (
    <Query query={USER_ORDERS_QUERY}>
      {({ data: { orders }, error, loading }) => {
        if (error) return <Error error={error} />;
        if (loading) return <p>Loading...</p>;
        return (
          <div>
            <h2>You have {orders.length} orders</h2>
            <OrderUl>
              {orders.map(order => (
                <OrderItemStyles key={order.id}>
                  <Link href={{ pathname: "/order", query: { id: order.id } }}>
                    <a>
                      <div className="order-meta">
                        <p>
                          {order.items.reduce((a, b) => a + b.quantity, 0)}{" "}
                          Items
                        </p>
                        <p>{order.items.length} Products</p>
                        <p>{formatDistance(order.createdAt, new Date())}</p>
                        <p>{formatMoney(order.total)}</p>
                      </div>
                      <div className="images">
                        {order.items.map(item => (
                          <img
                            src={item.image}
                            alt={item.title}
                            key={item.id}
                          />
                        ))}
                      </div>
                    </a>
                  </Link>
                </OrderItemStyles>
              ))}
            </OrderUl>
          </div>
        );
      }}
    </Query>
  );
};

export default Orders;
