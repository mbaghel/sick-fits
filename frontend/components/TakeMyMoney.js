import React from "react";
import StripeCheckout from "react-stripe-checkout";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";
import PropTypes from "prop-types";
import NProgress from "nprogress";
import Router from "next/router";
import Error from "./ErrorMessage";
import User, { CURRENT_USER_QUERY } from "./User";
import calcTotalPrice from "../lib/calcTotalPrice";

const CREATE_ORDER_MUTATION = gql`
  mutation createOrder($token: String!) {
    createOrder(token: $token) {
      id
      total
      charge
      items {
        id
        title
      }
    }
  }
`;

function tallyItems(cartItems) {
  return cartItems.reduce((tally, item) => tally + item.quantity, 0);
}

class TakeMyMoney extends React.Component {
  handleToken = async (token, createOrder) => {
    NProgress.start();
    console.log("handleToken called!");
    console.log(token.id);
    const order = await createOrder({ variables: { token: token.id } }).catch(
      err => {
        alert(err.message);
      }
    );
    Router.push({
      pathname: "/order",
      query: { id: order.data.createOrder.id }
    });
  };

  render() {
    return (
      <User>
        {({ data: { me } }) => (
          <Mutation
            mutation={CREATE_ORDER_MUTATION}
            refetchQueries={[{ query: CURRENT_USER_QUERY }]}
          >
            {createOrder => (
              <StripeCheckout
                amount={calcTotalPrice(me.cart)}
                name="Sick Fits"
                email={me.email}
                stripeKey="pk_test_SVVS6ApEXHjuHRTU3uMojtT0"
                description={`Order of ${tallyItems(me.cart)} items`}
                image={
                  me.cart.length && me.cart[0].item && me.cart[0].item.image
                }
                currency="USD"
                token={res => {
                  this.handleToken(res, createOrder);
                }}
              >
                {this.props.children}
              </StripeCheckout>
            )}
          </Mutation>
        )}
      </User>
    );
  }
}
export default TakeMyMoney;
