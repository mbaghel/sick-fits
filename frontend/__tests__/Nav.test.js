import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeUser } from "../lib/testUtils";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } }
  }
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } }
  }
];

describe("<Nav />", () => {
  it("renders a minimal nav when signed out", async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(toJSON(nav)).toMatchSnapshot();
  });

  it("renders the full nav when signed in", async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const nav = wrapper.find('ul[data-test="nav"]');
    const links = nav.find("Link");
    expect(toJSON(links)).toMatchSnapshot();
    expect(nav.exists("CartCount")).toBe(true);
    expect(nav.exists("Signout")).toBe(true);
  });
});
