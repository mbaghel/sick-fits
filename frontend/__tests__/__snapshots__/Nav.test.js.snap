// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders a minimal nav when signed out 1`] = `
<ul
  className="NavStyles-sc-11c0d2g-0 bLGDDS"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/signup"
  >
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign In
    </a>
  </Link>
</ul>
`;

exports[`<Nav /> renders the full nav when signed in 1`] = `
Array [
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>,
  <Link
    href="/sell"
  >
    <a
      href="/sell"
      onClick={[Function]}
    >
      Sell
    </a>
  </Link>,
  <Link
    href="/orders"
  >
    <a
      href="/orders"
      onClick={[Function]}
    >
      Orders
    </a>
  </Link>,
  <Link
    href="/me"
  >
    <a
      href="/me"
      onClick={[Function]}
    >
      Account
    </a>
  </Link>,
]
`;
